src/gpio.c

```diff
@@ -0,0 +1,31 @@
+#include <stdint.h>
+
+#include "gpio.h"
+
+#define SET_BITS_MASK   0x3U
+
+// function to set bits
+static void    set_bits_with_offset(volatile uint32_t *target, uint32_t offset, uint32_t value)
+{
+    uint32_t temp;
+
+    temp = *target & ~(SET_BITS_MASK << offset);
```

Comment:
nitpick: This function assumes that bit field is 2-bits long (not all fields are). This also entangles volatile registers in the mix, so you cannot reuse this function if you need set bits in some random u32.

A slightly better interface might be `uint32_t set_bits_with_offset(uint32_t field, unsigned int offset, unsigned int width, unit32_t value)`, so you can use it as `GPIOx->MODER = set_bits_with_offset(*GPIOx->MODER, GPIO_init->pin * 2, 2, GPIO_init->mode);`.

The function is also small enough to be `inline`.